src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -332,19 +347,49 @@ class BlockchainImpl(
                                           noEmptyAccount: Boolean = false): InMemoryWorldStateProxy =
     InMemoryWorldStateProxy(
       evmCodeStorage,
-      ReadOnlyNodeStorage(nodesKeyValueStorageFor(blockNumber)),
+      ReadOnlyNodeStorage(nodesKeyValueStorageFor(blockNumber, cachedNodeStorage)),
       accountStartNonce,
       (number: BigInt) => getBlockHeaderByNumber(number).map(_.hash),
       stateRootHash,
       noEmptyAccounts = false
     )
 
-  def nodesKeyValueStorageFor(blockNumber: Option[BigInt]): NodesKeyValueStorage =
-    PruningMode.nodesKeyValueStorage(pruningMode, nodeStorage)(blockNumber)
+  def nodesKeyValueStorageFor(blockNumber: Option[BigInt], storage: NodesStorage): NodesKeyValueStorage =
+    PruningMode.nodesKeyValueStorage(pruningMode, storage)(blockNumber)
 
-  def pruneState(blockNumber: BigInt): Unit = PruningMode.prune(pruningMode, blockNumber, nodeStorage)
+  def pruneState(blockNumber: BigInt): Unit = {
+    val blockToPrune = getBlockToPrune(blockNumber)
+    val currentBestSavedBlock = appStateStorage.getBestBlockNumber()
 
-  def rollbackStateChangesMadeByBlock(blockNumber: BigInt): Unit = PruningMode.rollback(pruningMode, blockNumber, nodeStorage)
+    if (blockToPrune <= currentBestSavedBlock){
+      PruningMode.prune(pruningMode, blockNumber, cachedNodeStorage, inMemory = false)
+    } else{
+      PruningMode.prune(pruningMode, blockNumber, cachedNodeStorage, inMemory = true)
+    }
+  }
+
+  def rollbackStateChangesMadeByBlock(blockNumber: BigInt): Unit ={
+    val currentBestSavedBlock = appStateStorage.getBestBlockNumber()
+
+    if (blockNumber <= currentBestSavedBlock)
+      PruningMode.rollback(pruningMode, blockNumber, cachedNodeStorage, inMemory = false)
```

Comment:
Should we clear the cache in this case?
If a node was deleted at a block `x` (it's new reference count is in cache) and we rollback to the block `x-1`, wouldn't we still have in cache the reference count after the delete? If that's so this would be eventually be persisted. WDYT?

Reply:
I think we are thinking of the same scenario, but just to double check, it would be the case that a rollback is done to block `best - 1` but the state of `best` is left in memory, so if persisted the state would become `best` (while the best block is now `best - 1`), right?

Reply:
yes, we would save this stale state to the disk, although I dont think it should generate more problems than just junk on disk. ( best - 1 state is still intact on disk) 
So when later when executing `newBest`  it would have all needed nodes and its snapshots would just overwrite old ones.

Reply:
:thinking: I'm not sure if that's the case, @AlanVerbner , won't persisting that state possibly wrongly increase or decrease the reference of a node?

Reply:
Saving junk nodes into DB is _not an issue_ The problem is when you mess around with snapshots as them are used to `undo` changes.
Let's say that you rollback between `[N, N+10]`. That means that we will replace by data in the snapshots for that range. If we rollback again between `[N+3, M], M > N + 10` and we have invalid data (snapshots that are no longer valid) we will undo with wrong data.

I need to re-read this specific case but it seems there might be some invalid snapshots in the cache, am I right @KonradStaniec @ntallar?

Reply:
@AlanVerbner if I understand correctly `rollback` do not generate snapshots, only all state changes to from given block? If that's the case there won't be stale snapshots, only duplicated state of the best block (from the diagram) which can be further persisted on the wrong branch.

Reply:
`rollback` does not generate snapshots. It uses them to replace data that's no longer valid.